### PR TITLE
Don't report a conflict between a function and a non-function

### LINKED ISSUE #143: Don't report a conflict for a function with same name as a non-function
I feel like #113 solves #112 but then goes a wee bit too far. Both the issue and the PR talk about *function* masking, which is really bad and confusing.

But I find I often have a name collision between
1. an unexported function name (e.g. `reprex:::preview_file()`)
2. an R object (e.g. a file path  `preview_file`) that exists somewhere in the body of a function and that, therefore, might conceivably exist in the global environment during development

This is a name collision between a function and a non-function and it doesn't seem to hurt anything, in the years I've used `load_all()` before #113.

But now, when this happens `load_all()` gives you a (IMO spurious) warning about a conflict. And it even implies that it's a function vs. function problem, even when it's not. Notice the parentheses following `all_named()` below in the warning about masking:

``` r
devtools::load_all("~/rrr/pkgload")
#> ℹ Loading pkgload

# pkgload has a function by this name
all_named <- FALSE

devtools::load_all("~/rrr/pkgload")
#> ℹ Loading pkgload
#> Warning: 
#> ── Conflicts ────────────────────────────────────────────── pkgload conflicts ──
#> x all_named() masks pkgload::all_named()
#> 
#> Did you accidentally source a file rather than using `load_all()`?
#> Run `rm(list = c("all_named"))` to remove the conflicts.
```

<sup>Created on 2021-02-23 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0.9002)</sup>

I think we should not warn for name collisions between a function and a non-function.

### PR DESCRIPTION
**Pull request by jennybc**
Closes #143

### CODE REVIEW

**Comment by jennybc** [comment]
Thinking about this more, it feels like the conflict messaging still unconditionally implies that a function is masking a function, without actually checking that this is true (e.g. they could be data objects, yeah?). I won't get into this here, but maybe it's worth its own issue.

**Comment by hadley** [comment]
Looks good to me!

**Comment by jimhester** [comment]
I think a simpler way to approach this might be to use something like this to get the exported functions, then keep the rest of the code as is.

```r
get_function_exports <- function(ns) {
  if (isNamespace(ns)) {
    nms <- getNamespaceExports(ns)
  } else {
    nms <- names(ns)
  }

  nms[exists(nms, ns, mode = "function")]
}
```

**Comment by hadley** [comment]
The downside would be not reporting conflicts between non-functions, but that's probably fine (although possibly confusing)

**Comment by jimhester** [comment]
If we do want to keep the existing logic `xor()` is vectorized, so if you convert to use `exists()` as I show above you can do away with the `vapply()`.

**Comment by jennybc** [comment]
Since we're refining, here's a demo of the data case, which is pretty easy to show with reprex right now. It recently gained an internal vector of file slugs.

``` r
devtools::load_all(".")
#> ℹ Loading reprex

# this is an internal data object of slugs
head(adjective_animal)
#> [1] "rude-stud"   "fussy-agama" "next-esok"   "legal-mink"  "alive-lark" 
#> [6] "smoky-macaw"

# mask it in globalenv
adjective_animal <- "sleepy-dog"

# note the spuriour function-y messaging
devtools::load_all(".")
#> ℹ Loading reprex
#> Warning: 
#> ── Conflicts ─────────────────────────────────────────────── reprex conflicts ──
#> x adjective_animal() masks reprex::adjective_animal()
#> 
#> Did you accidentally source a file rather than using `load_all()`?
#> Run `rm(list = c("adjective_animal"))` to remove the conflicts.
```

<sup>Created on 2021-02-24 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0.9002)</sup>

**Comment by jennybc** [comment]
Is it easy to access the fact we actually want, namely that one thing actually masks the other in practice?

**Comment by jimhester** [comment]
I generally think we could punt on the data masking data case, it seems more likely to happen by choice rather than by accident than the function case.

**Comment by jennybc** [comment]
It is certainly true that all my experience with unnecessary conflict reporting has been: a non-function object, e.g., a vector, in `globalenv()` "masking" (scare quotes intentional!) a function in the package namespace (which is artificially large in the `load_all()` scenario and replete with tempting names relevant to the package).

OK I'll take another pass as this today.

**Comment by hadley** [comment]
Also fixes #125

File: R/load.r [hunk 1]
```diff
@@ -258,6 +263,20 @@ warn_if_conflicts <- function(package, nms1, nms2) {
...
 
+get_function_exports <- function(ns) {
+  if (isNamespace(ns)) {
+    nms <- getNamespaceExports(ns)
+  } else {
+    nms <- names(ns)
+  }
+
+  if (length(nms) == 0) {
+    character()
```
Comment (jennybc) [inline R/load.r:274]: Without this, some S4-related tests fail

**Comment by jennybc** [comment]
> I think a simpler way to approach this might be to use something like this to get the exported function names, then keep the rest of the code as is.

I did this but now I notice different tests fail (re: bindings being forced).

File: R/load.r [hunk 2]
```diff
@@ -258,6 +263,22 @@ warn_if_conflicts <- function(package, nms1, nms2) {
...
+  } else {
+    is_fn <- vapply(nms, FUN.VALUE = logical(1), FUN = exists,
```
Comment (jennybc) [inline R/load.r:276]: Note that `exists()` is not vectorized.

**Comment by jimhester** [comment]
Thank you!

**Pull request merged**